### PR TITLE
Overhaul

### PR DESCRIPTION
**Pull request by nimmolo**
Hi Silvia - 

You will notice this is a pretty big recode - needs testing.
Please check it out and let me know what you think.

I've paid the most attention so far to 

- fixing the API calls 
- abstracting the plugin into classes
- abstracting all vendor-dependent functions into separate classes, so providers can be added more easily. the add-author forms are now special for each vendor, so the different vendors' names for required API tokens are not so confusing. the whole form is generated recursively from a new options array, $options['hosts']. 
- getting rid of all dependencies on vendor php API libraries. youtube alone was 20MB - and i figured if i can hardcode their crazy API calls without the library (they require 3 separate calls to establish authentication) i might as well hardcode all of the vendors. testing this took a lonnng time
- AJAXifying the admin and smoothing out admin UI

What i haven't paid as much attention to

- escaping, sanitizing, etc. I actually don't understand this very well and would appreciate advice
- dotsub embeds work, the iframe makes the layout strange
- namespacing and breaking out more classes so the main plugin file isn't so long!

### CODE REVIEW

File: readme.txt
Comment (silviapfeiffer) [inline readme.txt]: might want to add DotSub back in ;-)

Comment (nimmolo) [reply readme.txt]: :) yes and Dailymotion!

Comment (nimmolo) [reply readme.txt]: Just making screenshots now, and one more option to "add external-videos to the main post loop" which i use at the kiteboat site, might be useful for others

Comment (nimmolo) [reply readme.txt]: the media gallery hooks don't work anymore because of media using backbone.js, which i have to learn. for next release!

**Comment by silviapfeiffer** [comment]
Looks quite the massive patch! I do prefer smaller patches with more focused changes rather than one big one, but that might be a little late now. I'm going to have to check it out and give it a test.

**Comment by nimmolo** [comment]
i hear you. sorry -

yeah the apis and the AJAX were a massive learning curve for me, and took forever to troubleshoot. i figured if i was already spending weeks of work time on it, i might as well go all in. :) i think i put in like 150 hours on this, no joke. not that it should have taken that long, but it was a learning thing too.

**Comment by silviapfeiffer** [comment]
Just got it working (had to comment out the echo() function in ev-wistia). First look is really very impressive - I'm liking the new settings area very much! Now need to get API keys and stuff...

**Comment by silviapfeiffer** [comment]
A bit disappointed that it moved all my old ev-videos into trash despite still knowing the link from earlier. Trying to get them back now.

**Comment by nimmolo** [comment]
Going to sync a new commit in ~5 minutes, i think it fixes quite a few oversights like that

**Comment by silviapfeiffer** [comment]
OK, so the YouTube API setup page asks for a channel name. If entered incorrectly, it complains about wrong "author". Could be a be it confusing.

**Comment by nimmolo** [comment]
Got it - will fix. 

trying to sync with the different API terminology at each host, so users will know where to look for the info.

**Comment by silviapfeiffer** [comment]
In fact, I don't seem to be able to get past that problem. I keep getting "Invalid Author". :-(

**Comment by nimmolo** [comment]
hahaha i'm feeling like an "invalid author" now!!! 

just kidding i did successfully add youtube author locally with the new commit

please download my next commit, i think that is fixed. should be just a few minutes now

**Comment by silviapfeiffer** [comment]
Having the same trouble with vimeo - will wait until you've got it fixed. :-)

**Comment by nimmolo** [comment]
OK commit is in. Should fix most of those problems... 

wasn't just 5 minutes! was it? github is keeping me honest

**Comment by nimmolo** [comment]
btw... Dotsub embeds are just small, seems no way around it. I'd say the Dotsub owners are probably out "disrupting" some other part of the world, the whole project seems unloved at the moment. Documentation of the API is certainly full of ancient uncorrected errors.  

Anyway to fix Dotsub iframe layout, what's needed is some CSS on the theme side. 

Needs `.entry-video iframe { width: 100%; height: whatever; }`

**Comment by silviapfeiffer** [comment]
Hmm, I'm still getting "Invalid Channel Name" on YouTube.

**Comment by silviapfeiffer** [comment]
Undefined variable: channelsUrl in ../wordpress/external-videos/hosts/ev-youtube.php on line 84

**Comment by silviapfeiffer** [comment]
Turn that line into:
`$channelsSearch = wp_remote_get( $url );`
and YouTube starts working.

**Comment by silviapfeiffer** [comment]
You might want to check your debug.log in the wordpress/wp-content/ directory for errors - I have a whole bunch. I'm just reporting on those that stop things from working.

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: replace this with `if( isset($delete_messages) ) {`

File: core/ev-settings-forms.php
Comment (silviapfeiffer) [inline core/ev-settings-forms.php]: add `$ev_loop = $options['loop'];`

**Comment by silviapfeiffer** [comment]
OK, I got vimeo working also, phew. It's really hard to find that User ID, we might need to add an image to help people?

**Comment by silviapfeiffer** [comment]
Having trouble with getting videos from vimeo now:
`Undefined index: data in /Users/silviapfeiffer/src/wordpress/external-videos/hosts/ev-vimeo.php on line 216`
`Undefined index: paging in /Users/silviapfeiffer/src/wordpress/external-videos/hosts/ev-vimeo.php on line 217`

**Comment by silviapfeiffer** [comment]
OK, I seem to have had an error with the personal access token - we need better error notifications (I'll work on that).

**Comment by nimmolo** [comment]
whoo - sorry about all that! Was up late last night deleting and re-adding hosts to test and thought i cleared all the errors, but i kept monkeying around with the code obviously.

thanks - I made the line changes you noted above, but i'm going to delete and re-add all the accounts again. There is a youtube function i have in backup that is more forgiving, i may reinstate that. it looks for your channel by your _screen_ name which is the forward-facing version of the channel name. For kiteboat, for example, their channel says "kiteboat" to the world but the URL is "thekiteboatproject". Seems absurd that Google doesn't give you a straight-up User_ID.

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: Assuming there is an API error in one of the hosts, we don't get a proper return for the $new_videos, so this line is activated. However, this line avoids all error messaging to the user and just prints "0" on the interface - not a very helpful error message. It would be better to do a try{} catch{} on the fetch_new_videos call and catch an exception, and then inside the fetch calls in the individual sites we would throw an exception when things don't work.

You can test this with Vimeo and providing a broken personal access token.

Comment (nimmolo) [reply core/ev-admin.php]: Got it. will do

**Review by silviapfeiffer — CHANGES_REQUESTED** [review]
Looking generally good, but a few more things to do for stability.

**Comment by silviapfeiffer** [comment]
Hey, don't get me wrong, this is great work! I'm just ironing out kinks!

File: core/ev-admin.php
Comment (silviapfeiffer) [reply core/ev-admin.php]: In ev-vimeo.php, you might want look at what returns in $body and do something like this:
`
if ( $body['error'] ) {
  throw new Exception($body['error']);
}
`

**Comment by silviapfeiffer** [comment]
BTW: Dailymotion, Wistia and DotSub worked without a hitch! Nice work!

**Comment by nimmolo** [comment]
thanks thanks! i'm on it...

how does that change request work btw - anything there not in this thread? because there are like 21,000 diffs... this is my first time doing this

**Comment by silviapfeiffer** [comment]
What will happen is that eventually I'll be happy to apply your change to my master branch. Then I'll do some more final changes and make a release.

This is indeed a very big change request - normally they are no more than a couple of dozen lines. I would normally expect the changes to the individual sites each to be in their own pull request (particularly the removal of the Google SDK). But it's ok, I can deal with it.

There's currently nothing else waiting to be released so this will be the next big release.

**Comment by nimmolo** [comment]
ok sounds great. it will probably be tomorrow your time when i make that commit (currently 7:27pm here pacific time usa). or is it already tomorrow? you're in germany right?

**Comment by nimmolo** [comment]
HA! Sorry Australia! It says so right in your profile. Somehow I thought you were in Germany ever since.

File: hosts/ev-wistia.php
Comment (silviapfeiffer) [inline hosts/ev-wistia.php]: large_size_w and large_size_h are not currently defined - these should be hard coded to 180x135 (also note that the "X" below is capitalised, so creates an error with wistia.

File: hosts/ev-wistia.php
Comment (silviapfeiffer) [inline hosts/ev-wistia.php]: replace capital "X" with small "x"

**Comment by silviapfeiffer** [comment]
I am a German living in Australia - any mixups are completely understandable! Have a nice evening and keep going - this is going to be a very nice new release!

**Comment by nimmolo** [comment]
Hi Silvia, I think i'm done. 

* No php or js errors in my testing the last hour
* On Vimeo "add author" form, changed label of "User ID" to "User ID Number" to make that extra clear, plus clarified "explanation" field. URL of page where this is available included also.
* Cleared all the API errors with google. An oddity of Google is that the username... is not necessarily the username. Kiteboat's front-facing username is "kiteboat", however an api search with parameter "forUsername=kiteboat" will not get you anything at YouTube, because their username is permanently set at "thekiteboatproject". I reverted to the "search for channel by keyword" method suggested by YouTube API docs and explained in the comments for ev-youtube.php - I just couldn't believe this was the most direct way... but it is.
* Fixed that mistake that was trashing other channels' videos when you update a channel
* Fixed various js bugs

**Comment by nimmolo** [comment]
Hope you enjoy the icon. My girlfriend made it - it's kind of based on an old photo of the Parkes Observatory in Australia, crossed with the more prosaic rooftop dish.

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: Hmm this was there to deal with multiple messages created in this function. Why would we not need that any longer?

**Comment by silviapfeiffer** [comment]
I think there's a logic error somewhere: now, when you press the "update videos from all channels" button, it deletes all videos.

Related error messages:

`Undefined variable: video_ids in external-videos/core/ev-admin.php on line 386`
`in_array() expects parameter 2 to be array, null given in external-videos/core/ev-admin.php on line 386`
`Undefined variable: messages in external-videos/core/ev-admin.php on line 404`

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: Hmm, you've circumvented the problem with error messages from the hosts by simply saying "No videos found" rather than displaying the error reported, e.g. on Vimeo you could get this as a reply: 

``Array
(
    [error] => You must provide a valid authenticated access token.
)``

but we now simply say "No videos found". I think we can handle this better.

For example, I had suggested creating exceptions in ev-vimeo.php when an error is returned and handling the exception in ev-admin.php.

If I'm not explaining this well enough, I can fix that up after merging, but I'd be happy for you to give it a go.

**Comment by silviapfeiffer** [comment]
The telescope icon is nice, but to be honest, I fail to understand where it's being used in the plugin.

Once you've fixed up the "update videos from all channels" button, I'm prepared to merge. It's looking really good! :-)

**Comment by nimmolo** [comment]
Embarrassing!   One more commit coming after i take a break. 

the icon is just for the WP plugin repo, otherwise it gets assigned one of the "texture" icons

**Comment by silviapfeiffer** [comment]
Oh, I had no idea about the repo icon! Nice one!

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: is this all commented out?

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: ??

**Comment by silviapfeiffer** [comment]
I'm pretty happy with what you're doing, so don't apologize!

One tip: your commit messages should explain what the commit fixes. Write a full sentence to explain what problem you are addressing and how. That makes reviewing so much easier. For example: that last commit - what did you fix with it?

File: core/ev-admin.php
Comment (nimmolo) [reply core/ev-admin.php]: no - those are commented-out comment-out marks.
they're easier to turn off and on that way, at least in my text editor (atom)
removing them now!

File: core/ev-admin.php
Comment (nimmolo) [reply core/ev-admin.php]: removing this too. 

the nice thing about this style of commenting-out is that if you do this line/block comment at both the opening `// /*` and the closing ` // */ ` you can turn the block of code on and off without altering the closing line. faster.

at least in the text editor i'm using, atom, you only toggle the top block-comment symbol  `/*` on and off using `command-/`, which adds the line comment `//` before it, and disables the block comment (preserving the active code). but if you un-comment the block comment symbol, the closing `// */` will terminate any block comment, whether or not there is one, without affecting code. good for testing.

if that's confusing disregard. probably makes more sense in practice, if you have a similar text editor.

Comment (silviapfeiffer) [reply core/ev-admin.php]: Yeah, I knew they were related - just looked like some left over debugging code. Not to worry! :-)

**Comment by nimmolo** [comment]
Silvia - there was logic problem indeed, in the update/delete videos loop - it looked for the added_videos result one dimension up from the array where it was stored. this caused the function to always find all remote videos deleted, even during an update, and so everything kept getting trashed. the last commit fixed that.

**Comment by nimmolo** [comment]
Something occurred to me: if users have some channels already established with videos imported, but without API keys, they would currently have to delete their channels and re-add them to make the channels work again, losing their videos. so I'm making little AJAX forms to edit existing users now. :)

**Comment by silviapfeiffer** [comment]
Awesome!

**Comment by nimmolo** [comment]
If a user has the same username at different hosts, this could result in duplicate "author_id"s in the authors table. we were working around this by double checking the author's "host_id". 

However, author forms are fully customized by host now, and we're always only checking for authors in a host context. If the author options were to be broken up and indexed by host, i found it really simplifies all functions related to looping them, checking authors and customizing host author forms. An author is really an account at a host, in other words, and is thus always a property of that host.

This new change makes that db change. No authors are discarded, but an init function moves each $author under $options['hosts'][$host_id]['authors'][$author_id] and unsets $options['authors']

**Comment by silviapfeiffer** [comment]
Wow, you've worked hard on this. I'll do some thorough testing and see if we're ready to land.

**Comment by nimmolo** [comment]
I caught an error in the cron job overnight... fixed

**Comment by nimmolo** [comment]
Hi Silvia,
Happy new year! Hope you're doing well, especially if you had to go back to work.

I found a couple more errors that I'll be updating in a commit today. (cron job and vimeo embed url)

But also wanted to tell you, I'm on the verge of cracking the extensions to Media Backbone.js that allow browsing and placing video from a user's External Videos channels into post content. This is a feature that's been broken since 3.8 or so. I'm pretty excited about it - it's really been a bear to figure out.

screen shot 2017-01-06 at 2 53 29 pm

I don't know if you want to bring this feature in with this release or in a 1.1 release, let me know if so.

**Review by silviapfeiffer — APPROVED** [review]

**Comment by silviapfeiffer** [comment]
I'll do some testing today and maybe do a release now, so your next fixes can land with the next release. No point in holding more things back.

**Comment by silviapfeiffer** [comment]
FYI: I'm staring at the options handling and I can't help but wanting to clean it up. We used to have no class and therefore we used to load the options every time we needed them and then save any changes. Now with a class around, we should only load the options once via get_option('sp_external_videos_options') and only touch the stored options again when a change is necessary.
E.g. I've already noticed that the 'slug' option keeps getting lost because it's handled differently from the rest.

I'm adding a change to this PR for it so you can have a look before I land it.

**Comment by nimmolo** [comment]
I like the sound of that!

I wondered about the number of times i was "getting" the options. Maybe it's really simple to clean up, but i'm looking forward to seeing how you do it.

**Comment by silviapfeiffer** [comment]
https://github.com/silviapfeiffer/external-videos/commit/728725bfd56b819328ea0e0f3a7cbb4f68e8bf53 - my options handling cleanup - less than I thought because doing an in-memory handling across the different classes would have been too big a change.

**Comment by silviapfeiffer** [comment]
I think the cleanest approach would be to create a separate class that handles both the reading and the updating of options and that the other classes interface with. Haven't got the time to do it, but if you're keen, go ahead.
I'm moving on to testing other things. :-)

**Comment by silviapfeiffer** [comment]
BTW: "Update videos from all channels" is broken again :-(

**Comment by silviapfeiffer** [comment]
Definitely something funky going on with:
  function post_new_videos( $update_authors, $update_hosts, $single )
.. the "$single" parameter isn't even used

File: core/ev-admin.php
Comment (silviapfeiffer) [inline core/ev-admin.php]: get_authors is used by both this function and the "Update videos from all channels" button in the settings. I'm preparing a fix for get_authors rather than doing this manually.

**Comment by silviapfeiffer** [comment]
https://github.com/silviapfeiffer/external-videos/commit/905b17781c8dd47e46871c72180745e5e26703b3 fixes the problem with multiple authors

**Comment by silviapfeiffer** [comment]
I also have a fix for not deleting other site's videos when checking for new updates: https://github.com/silviapfeiffer/external-videos/commit/4433e84d80a1510d3d65d2838237b4bf085ff529

**Comment by silviapfeiffer** [comment]
I'm getting close to landing these patches - you can then add any further functionality onto a new branch, if that's ok.

**Comment by silviapfeiffer** [comment]
OK, I've landed it on master, but I've found that we're missing the ability to remove authors again. I'll hold off on the release until we've re-added that function. I'll put an issue in for this also.

**Comment by nimmolo** [comment]
Thanks for going through it. Just downloaded it - seems like the js is broken... I'm getting all the tabs loaded at once, and existing authors are all gone.

One issue you mentioned I definitely fixed a few days ago: deleting other sites videos when checking for new updates. Just to check, for sure you were working with the recent commits?

re: the options class. It sounds like you understand how classes interact better than i do - i'm all ears if you can explain the principle involved there!

**Comment by nimmolo** [comment]
I wasn't able to find the source of the js problem, so i'm working forward from my most recent development version, and adding in your most recent changes mentioned in this thread, one by one. 

One thing I now understand, that you noticed too, is the faulty logic in my fetch_new_videos and post_new_videos functions. But the deeper and more confusing problem (which i had been mistakenly trying to solve by passing the $single variable) was my still using an authors loop rather than a hosts loop in fetch_new_videos. This function really only has two modes: it updates one host/author, or it updates all hosts/all authors. I've rewritten the function to pass an $update_host and an $update_author, or $update_hosts and null, which means do all authors. 

Oh and I FINALLY also understood something you've been saying all along :-D which is that my function trashed deleted videos for EVERY author, even if you were only wanting to update a single author. I've enclosed the trash_deleted_videos within the update-all conditional so it only happens if you're updating all, or running the cron job.

If you'll give me brief directions about abstracting the options into a class, i can tackle this too.

**Comment by nimmolo** [comment]
I wasn't able to find the source of the js problem, so i'm working forward from my most recent development version, and adding in your most recent changes mentioned in this thread, one by one.

One thing I now understand, that you noticed too, is the faulty logic in my fetch_new_videos and post_new_videos functions. But the deeper and more confusing problem (which i had been mistakenly trying to solve by passing the $single variable) was my still using an authors loop rather than a hosts loop in fetch_new_videos. This function really only has two modes: it updates one host/one author, or it updates all hosts/all authors. I've rewritten the function to pass an $update_host and an $update_author, or $update_hosts and null, which means do all authors.

This allows me to eliminate the confusing function get_authors, which i should have done already.

Also: I FINALLY also understood something you've been saying all along, which is that my function trashed deleted videos for other authors, even if you were only wanting to update a single author. The problem here is a bit deeper than we thought. The trash_deleted function was checking the array of $new_video_ids, against the array of $all_videos (all existing EV videos). But if we're only updating one author, all other authors' videos will always be missing from this array. So I'm changing the trash_deleted function to check the array only against videos by that same (host, author) if we're updating one author only. If we're updating all authors, it uses the whole array as before.

If you'll give me brief directions about abstracting the options into a class, i can tackle this too.

**Comment by silviapfeiffer** [comment]
Please restart a new branch from my master branch. Many of the things you're talking about have already been fixed and landed.

**Comment by silviapfeiffer** [comment]
Also, from now on let's just make one pull request per issue. First you register the issue and then you provide a PR to address the issue. That way it's actually possible for me to review if your PR fixes a particular problem.

**Comment by nimmolo** [comment]
OK - understood! 

My note was because i downloaded the master branch, and it seems not to work ... is master branch up to date with your fixes? I'm getting broken settings screen

**Comment by silviapfeiffer** [comment]
Oh! I din't notice - let me check. Just as well I haven't managed to publish yet (SVN is SOOO slow).

**Comment by silviapfeiffer** [comment]
Bugger, not sure what I did last night - I seemed to work just fine. Let me fix that.

**Comment by silviapfeiffer** [comment]
something about ev-admin.css not loading...

**Comment by silviapfeiffer** [comment]
Strangely it works on my local branch, but not on a new re-checked out from master. Also, all my commits are on master and diff doesn't show any file differences. I have to go to work, but will take another look tonight.

Thanks for working forward from your existing working branch! Doing a bisect is all I can think of right now also.

**Comment by nimmolo** [comment]
ok. sorry - what's a bisect? i am truly ignorant of the terminology and procedures in collaborative development. 

i can have a version ready for you to test tonight, with your changes pulled in from the diff, plus what i believe to be much improved logic in post_new_videos, fetch_new_videos, and trash_deleted_videos, if you let me know where to put it.

**Comment by silviapfeiffer** [comment]
Bisect is what you're doing: finding the patch that broke the settings page by walking through the past patches. 

Once you know what to fix, create a new branch from master and apply your patches on there. Then create a new pull request (or several if they fix different issues). You can branch off your fixed branch also for a new pull request. 

I have a gut feeling that some JS is broken somewhere and fails silently in the browser. Then again, I touched very little JS yesterday...

**Comment by nimmolo** [comment]
thanks re: bisect.

re: js, that is odd - it's not showing that there were any diffs in your ev-admin.js file!

**Comment by nimmolo** [comment]
ok - I branched off my fixed branch, with all your commits pulled in (minus a few conflicts). Explanation in new pull request.

**Pull request merged**
